File: packages/java/src/main/java/ai/alumnium/driver/BaseDriver.java
package ai.alumnium.driver;

import ai.alumnium.accessibility.BaseAccessibilityTree;
import ai.alumnium.tool.BaseTool;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Set;

/**
 * Abstract surface every driver flavour must provide.
 *
 * <p>Sealed on the three supported backends - Selenium, Playwright, Appium - so that {@code switch}
 * pattern matching on {@code BaseDriver} is exhaustive.
 */
public abstract sealed class BaseDriver permits SeleniumDriver, PlaywrightDriver, AppiumDriver {

  public abstract String platform();

  protected abstract BaseAccessibilityTree fetchAccessibilityTree();

  private BaseAccessibilityTree cachedAccessibilityTree;

  public BaseAccessibilityTree accessibilityTree() {
    if (cachedAccessibilityTree == null) {
      setAccessibilityTree(fetchAccessibilityTree());
    }
    return cachedAccessibilityTree;
  }

  public void setAccessibilityTree(BaseAccessibilityTree tree) {
    cachedAccessibilityTree = tree;
  }

  public void resetAccessibilityTree() {
    cachedAccessibilityTree = null;
  }

  public abstract Set<Class<? extends BaseTool>> supportedTools();

  public abstract void click(int id);

  public abstract void dragSlider(int id, double value);

  public abstract void dragAndDrop(int fromId, int toId);

  public abstract void pressKey(Key key);

  public abstract void quit();

  public abstract void back();

  public abstract void visit(String url);

  public abstract String screenshot();

  public abstract void scrollTo(int id);

  public abstract String title();

  public abstract void type(int id, String text);

  public abstract String url();

  public abstract String app();

  public abstract Object findElement(int id);

  public abstract void executeScript(String script);

  public abstract void switchToNextTab();

  public abstract void switchToPreviousTab();

  public abstract void printToPdf(String filepath);

  /** Hover is only meaningful for web drivers; mobile throws by default. */
  public void hover(int id) {
    throw new UnsupportedOperationException("hover is not supported by " + platform());
  }

  /** File upload is only meaningful for web drivers; mobile throws by default. */
  public void upload(int id, List<String> paths) {
    throw new UnsupportedOperationException("upload is not supported by " + platform());
  }

  /** Loads a classpath-shipped script (e.g. waiter.js) as a UTF-8 string. */
  protected static String loadScript(String resourcePath) {
    try (InputStream in = BaseDriver.class.getResourceAsStream(resourcePath)) {
      if (in == null) {
        throw new IllegalStateException("Missing driver script resource: " + resourcePath);
      }
      return new String(in.readAllBytes(), StandardCharsets.UTF_8);
    } catch (IOException e) {
      throw new IllegalStateException("Failed to read driver script resource: " + resourcePath, e);
    }
  }
}

File: packages/python/src/alumnium/drivers/base_driver.py
from abc import ABC, abstractmethod

from ..accessibility import BaseAccessibilityTree
from . import Element
from .keys import Key


class BaseDriver(ABC):
    @property
    def accessibility_tree(self) -> BaseAccessibilityTree:
        cached = getattr(self, "_cached_accessibility_tree", None)
        if cached is None:
            cached = self._fetch_accessibility_tree()
            self.set_accessibility_tree(cached)
        return cached

    def set_accessibility_tree(self, tree: BaseAccessibilityTree):
        self._cached_accessibility_tree = tree

    def reset_accessibility_tree(self):
        self._cached_accessibility_tree = None

    @abstractmethod
    def _fetch_accessibility_tree(self) -> BaseAccessibilityTree:
        pass

    @abstractmethod
    def click(self, id: int):
        pass

    @abstractmethod
    def drag_slider(self, id: int, value: float):
        pass

    @abstractmethod
    def drag_and_drop(self, from_id: int, to_id: int):
        pass

    @abstractmethod
    def press_key(self, key: Key):
        pass

    @abstractmethod
    def quit(self):
        pass

    @abstractmethod
    def back(self):
        pass

    @abstractmethod
    def visit(self, url: str):
        pass

    @property
    @abstractmethod
    def screenshot(self) -> str:
        pass

    @abstractmethod
    def scroll_to(self, id: int):
        pass

    @property
    @abstractmethod
    def title(self) -> str:
        pass

    @abstractmethod
    def type(self, id: int, text: str):
        pass

    @property
    @abstractmethod
    def url(self) -> str:
        pass

    @property
    @abstractmethod
    def app(self) -> str:
        pass

    @abstractmethod
    def find_element(self, id: int) -> Element:
        pass

    @abstractmethod
    def execute_script(self, script: str):
        pass

    @abstractmethod
    def switch_to_next_tab(self):
        pass

    @abstractmethod
    def switch_to_previous_tab(self):
        pass

    @abstractmethod
    def print_to_pdf(self, filepath: str):
        pass
